File: c++/leetcode/199. Binary Tree Right Side View.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ans; 
        
        if(!root) return ans;
        
        queue<TreeNode*> q;
        q.emplace(root); 
        
        
        while(!q.empty()) {
            const auto sz = q.size();
            
            for(int i = 0; i < sz; ++i) {
                const auto front = q.front();
                
                if(front -> left)
                    q.push(front -> left);
                
                if(front -> right)
                    q.push(front -> right);
                
                if(i == sz - 1)
                    ans.push_back(q.front() -> val);
                
                q.pop();
            }
            
        }
        
        return ans;
    }
};

File: javascript/leetcode/199. Binary Tree Right Side View.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var rightSideView = function(root) {
    const ans = []
    
    if(!root) return ans
    
    let q = [root]
    
    while(q.length) {
        const sz = q.length
        
        for(let i = 0; i < sz; ++i) {
            const front = q[0]
            
            if(front.left)
                q.push(front.left)
            
            if(front.right)
                q.push(front.right)
            
            if(i === sz - 1)
                ans.push(q[0].val)
            
            q = q.slice(1)
        }
    }
    
    return ans
};
